sharktank/sharktank/types/pipelining.py

```diff
@@ -30,34 +27,24 @@ def transfer_between_blocks_if_needed(
     Args:
         x: The input tensor to process.
         curr_block: The index of the current block.
-        block_to_pipeline_stage: A list mapping each block to its corresponding pipeline stage.
-        pipline_stage_to_devices: A list mapping each pipeline stage to the devices it uses.
+        theta: The theta object used
 
     Returns:
         The input tensor, possibly moved to different devices.
     """
-    assert (block_to_pipeline_stage is None) == (pipline_stage_to_devices is None), (
-        "Either both block_to_pipeline_stage and pipline_stage_to_devices are None, "
-        "or both are not None."
-    )
-
-    # No pipeline parallelism, nothing to do.
-    if block_to_pipeline_stage is None:
+    # Unsharded tensors do not need to be moved.
+    if not isinstance(x, ShardedTensor):
```

Comment:
Perform the transfer on unsharded versions too (it should be useful for replicated tensors as we don't need to handle transfers separately).

Reply:
So some extra changes would be needed - namely move to new device always transfers. For the initial PP we can have the redundant transfers of deviceN -> deviceN then rely on the compiler to clean them up. Overall we should using information that shouldn't be actually required (e.g. where its coming from).

Without tensor parallelism we are less worried about device placement confusion.